rfcs/013 - Datastore Revisited RFC.md

```diff
@@ -0,0 +1,137 @@
+
+### Minimally-invasive with restored type introspection
+
+We recently removed the ability to get the type of a variable as a string with the `type variableName` syntax, but if we restore that, it should be possible to force type safety by combining the type string with either the namespace or the key string, such that trying to get the same key name with a default value of a different type will find that it doesn't exist and return the provided default.
+
+This could provide zero actual API call changes except to the `del` function, which would need to be given an example of the type of data to delete in order to find it. This has the smallest change requirement, though it does require bringing back the string-based type manipulation code.
+
+### Generic Functions (in addition to the interface functions that currently exist)
+
+The maximum contrast to the prior is this particularly invasive change, where functions gain the ability to have one or more generic types, eg `fn foo<A>(arg1: int64, arg2: string): A {` such that they can be called with some type that is passed in during invocation but is not passed in as a variable, just the type, so `foo<bool>(1, 'bar')` would do something internally with the type.
+
+In this case, the `namespace` function would take not only the name parameter, but the type info for the kind of variable to store in it, like `namespace<Array<int64>>('variousDigitsOfPi')`, then this type would internally be applied to the namespace through some syntax similar to the `type variableName` before so it can be stored in the datastore opcodes, but it would also be used to make sure writes and reads are sound, always providing values of the correct type.
+
+The biggest issue here is: it adds a new "color" of function, and there will be pressure to add something like an `fn foo<A as Orderable>(...` syntax so the generics defined for the function are also interface constrained so the function knows what to do with it besides pass it around or poke at it in some way or the other.
+
+It'll make the language much more complex, but will the utility gain be worth the complexity? It would make constructor functions much simpler, eg instead of `newHashMap(firstKey, firstVal)` we get `new<HashMap<KeyType, ValType>>()` without requiring the first key-val pair to be defined.
+
+### Compiler-level hackery
+
+We could do it with neither of the above via effectively hackery within the compiler; the `ds*` opcodes get a special behavior that always appends the type information to the `namespace` string no matter what and we just have some magic behavior that "just works" but doesn't trust the end user to have access to that power.
+
+I'm including this for completeness, but that is the "Go Way" and in my book "The Go Way Is Stupid."
+
+### Rust-style impl-ish functionality
```

Comment:
IMHO, this would cause the syntax of Alan to become a *lot* more noisy. I primarily see associated functions as a great way to solve namespacing functions (instead of having to do `newHashMap`), but with function shadowing I don't see the need for it (ie, `toString(5)` and `toString(true)` will always resolve correctly).